modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/ExoPlayerDataSourceFactory.kt

```diff
@@ -104,20 +89,63 @@ class ExoPlayerDataSourceFactory @Inject constructor(
         )
 
         val dataFactory = cacheFactory ?: defaultFactory
+
         return when {
             episodeLocation.episode.isHLS -> HlsMediaSource.Factory(dataFactory)
             (clipRange != null) -> DefaultMediaSourceFactory(dataFactory, extractorsFactory)
             else -> ProgressiveMediaSource.Factory(dataFactory, extractorsFactory)
         }.createMediaSource(mediaItem)
     }
 
+    fun cacheDataSourceFactoryForStreamingUrl(url: String): CacheDataSource.Factory {
+        return cacheDataSourceFactory(defaultDataSourceFactory(uri = url, downloading = true))
+    }
+
+    private fun cacheDataSourceFactory(defaultFactory: DefaultDataSource.Factory): CacheDataSource.Factory {
+        return CacheDataSource.Factory()
+            .setUpstreamDataSourceFactory(defaultFactory)
+            .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
+            .let { if (cache != null) it.setCache(cache) else it }
+    }
+
+    private fun defaultDataSourceFactory(episodeLocation: EpisodeLocation): DefaultDataSource.Factory {
+        return defaultDataSourceFactory(uri = episodeLocation.uri, downloading = episodeLocation is EpisodeLocation.Stream)
+    }
+
+    private fun defaultDataSourceFactory(uri: String?, downloading: Boolean): DefaultDataSource.Factory {
+        val httpFactory = if (FeatureFlag.isEnabled(Feature.EXO_OKHTTP)) {
+            OkHttpDataSource.Factory(client)
+                .setUserAgent(Settings.USER_AGENT_POCKETCASTS_SERVER)
+                .also {
+                    addBasicAuthIfNeeded(uri = uri, downloading = downloading, it)
```

Comment:
I couldn't find a better way to add the HTTP header "Authorization". Hopefully it isn't expensive to create a new factory.

Reply:
Doing this on the Interceptor level is better. We can then use the same mechanism both for downloads and for streaming.